### DISCUSSION TITLE
Is JSON expression just LISP in JSON format?

### DISCUSSION THREAD

**Discussion by danielo515** [streamich/json-joy#674] [Q&A] [answered]
I just read the spec of JSON expressions https://jsonjoy.com/specs/json-expression and to me that looks exactly like any other LISP with a slightly twist on the syntax. So I thought, better ask. How is this different to LISP? Did you had LISP in mind when you designed this? I am just curious, not gonna suggest any change or anything.

**Comment by streamich** [comment, marked as answer]
Yes, it has the LISP-like syntax! I was actually also thinking to create LISP-like bracket grammar for is. So instead of JSON:

```js
[".", "a", "b"] // 'ab'
```

One could also write/read LISP-like debugging format:

```
(. "a" "b")
```

---

What I had in mind is to make it portable, compact, and fast.

- Portable &mdash; all modern programming languages support JSON.
- Compact &mdash; it consumes little space, can be embedded into JWT tokens, for example, for any authorization logic.
- Fast &mdash; `json-joy` complies JSON Expression to native JavaScript. For example, `['+', 1, 2]` complies directly to `() => 3`. If there is an external input in the expression, say `['+', 1, ['$', '/data']]`, it still complies to something fast like `() => 1 + io('data')`.

---

Regarding the syntax. JSON Expression is an S-expression which uses valid JSON syntax to define the expressions. It also uses JSON as its only data type, and also uses JSON as its only valid literals.

Any JSON Expression is a valid JSON value. For example, `1` is just `1`. Objects are just objects: `{"foo": "bar"}` is just `{"foo": "bar"}`. And so on.

The only exception from JSON is that arrays are treated as operators. So to represent a JSON array value, one has to use a [nullary operator](https://jsonjoy.com/specs/json-expression/operators-and-operands) for it, i.e. wrap it in extra `[ ... ]`. For example, a JSON array like `[1, 2, 3]` is represented in JSON Expresion as `[[1, 2, 3]]`.

**Reply by streamich** [reply]
Here is a [draft of what could be JSON Expression grammar](https://gist.github.com/streamich/09eb1683080f353df1943efa76b6863e).

**Reply by danielo515** [reply]
That looks very interesting, thanks
